### ISSUE TITLE
decorateReply hook does not match overload call

### ISSUE THREAD

**Issue by xirayah** [fastify/fastify#5285]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.25.2

### Node.js version

20.10.0

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

14.3

### Description

Fastify decorateReply hook no longer matches overload call of string and function with no parameters which was correct in previous versions and is still mentioned as working in reference documentation on fastify.dev site.

Fastify's reference documentation mentions below code as valid example of decorateReply hook

```
fastify.decorateReply('utility', function () {
  // Something very useful
})
```

I have checked the node_modules/fastify/types/instance.d.ts file between current version and the one I have been using previously (4.15.0).
The type for decorateReply has been changed and that's what is causing the issue.

### Steps to Reproduce

```
async function plugin(fastify: FastifyInstance, opts: FastifyPluginOptions) {
  // The below function() will be underlined in red with error saying that no overload matches this calll
  fastify.decorateReply('noParam', function() {
    this.status(400)
    this.send(MSG_NO_PARAM)
  })
  // This exact code works perfectly fine in 4.15.0
}
```

**Comment by climba03003** [comment]
Can you provide a reproducible repo?
It is working on my side, maybe try to update `typescript` version.
![image](https://github.com/fastify/fastify/assets/23028015/d4e9b303-35b8-45a2-b793-aad27b77e9a3)

**Comment by xirayah** [comment]
Thanks for poiting it out, I have already updated all of the packages on the repo today.

But now that you have mentioned that it works on another instance I have decided to investigate my repository again and I have managed to find the culprit.

In order to access the property on the reply parameter in functions I had to declare noParam in a module.
This the actual full code, I have overlooked the bottom part accidentally because it's in a shared file with other functions related to route specific utilities

```
async function plugin(fastify: FastifyInstance, opts: FastifyPluginOptions) {
  // This will decorate FastifyReply with .noParam
  fastify.decorateReply('noParam', function() {
    this.status(400)
    this.send(MSG_NO_PARAM)
  })
}

export default fp(plugin)

declare module 'fastify' {
  export interface FastifyReply {
    noParam: () => never
  }
}
```

All I had to do was change the
```
noParam: () => never
```
to
```
noParam:() => any
```
I am not even sure why I declared the function return as never in the first place.

Thanks for pointing me in the right direction 🙏
